Binary trees

Yegor932221 opened the pull request.

Commit:
Исходный код с практики

Commit:
Добавил функции isEmpty, Hight и NodeCount

Commit:
Переделал функцию поиска по ключу и исправил фунцию подсчёта узлов

Commit:
Написал функции максимума и минимума

Commit:
Добавил функции: проверки на сбалансированность, получение уровня узла,получения вектора ключей

Commit:
Добавил функции нахождения предка и дописал функцию удаления узла по ключу

tvz-041 commented:
Добавьте класс тестера, который мы писали на практике, и прогоните свой класс через него.

tvz-041 commented on binary trees/BinaryTree.cpp:
```diff
@@ -0,0 +1,515 @@
+	return Hight(m_root);
+}
+
+int BinaryTree::Hight(Node* root) const
```
```suggestion
int BinaryTree::Height(Node* root) const

```

Commit:
Исправил основной класс по коментариям

Commit:
Добавил тестер но не исправил ошибки

Commit:
Прогнал через тестер, исправил ошибки и добавил нужные функции

Исправил конструктор копирования, функцию клона, подсчёт количества нод, переписал удаление, добавил недостоющие const, исправил нахождение родителей и добавил фунции вывода и получения листов

tvz-041 reviewed (changes requested):
Прошу удалить списанные методы и реализовать их самостоятельно.

Commit:
Переписал функции

Commit:
Переписал remove, но иногда ошибка с переполнением в NodeCount

Commit:
Исправил ошибки

Я ошибся проблема выходила из-за того что проверка на корень в remove была после проверки на узел с двумя потомками из-за чего происходило обращение к родителю корня и ошибка из-за nullptr
